lib/auth/server_info.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/internalutils/stream"
+	"github.com/gravitational/teleport/api/types"
+)
+
+// ReconcileServerInfos periodically reconciles the labels of ServerInfo
+// resources with their corresponding Teleport SSH servers.
+func (s *Server) ReconcileServerInfos() error {
+	ctx := s.CloseContext()
+	const batchSize = 100
+
+	for {
+		// Iterate over nodes in batches.
+		nodeStream := s.StreamNodes(ctx, defaults.Namespace)
+		var nodes []types.Server
+		moreNodes := true
+		for moreNodes {
+			nodes, moreNodes = stream.Take(nodeStream, batchSize)
+			if err := s.setCloudLabelsOnNodes(ctx, nodes); err != nil {
+				return trace.Wrap(err)
+			}
+			select {
+			case <-s.clock.After(time.Second):
+			case <-ctx.Done():
+				return trace.Wrap(ctx.Err())
+			}
+		}
+	}
+}
+
+func (s *Server) setCloudLabelsOnNodes(ctx context.Context, nodes []types.Server) error {
+	for _, node := range nodes {
+		meta := node.GetCloudMetadata()
+		if meta != nil && meta.AWS != nil {
+			si, err := s.GetServerInfo(ctx, types.SubKindCloudInfo, meta.AWS.GetServerInfoName())
+			if err == nil {
+				if err := s.updateLabelsOnNode(ctx, node, si); err != nil {
+					return trace.Wrap(err)
+				}
+			} else if !trace.IsNotFound(err) {
+				return trace.Wrap(err)
+			}
+		}
+	}
+	return nil
+}
+
+func (s *Server) updateLabelsOnNode(ctx context.Context, node types.Server, si types.ServerInfo) error {
+	err := s.UpdateLabels(ctx, proto.InventoryUpdateLabelsRequest{
+		ServerID: node.GetName(),
+		Kind:     proto.LabelUpdateKind_SSHServer,
+		Labels:   si.GetStaticLabels(),
+	})
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		log.WithError(err).Debugf("no control stream for server %v", node.GetName())
```

Comment:
nit: In clusters with thousands of nodes, this log line would be extremely verbose.  Since we expect there to always be lots of nodes that are connected to other auths, I'd suggest instead aggregating a total count of the number of nodes that couldn't be updated due to not being connected and just log the total number once per periodic.